### DISCUSSION TITLE
left hand pose

### DISCUSSION THREAD

**Discussion by jimingda** [pattacini/study-air-hockey#4] [Q&A] [answered]
Does the initial pose(below) of study-1.cpp, generated from rotating the original xyz of left hand around its x axis by 270 or 90 degree?
```c++
        yarp::sig::Matrix R = yarp::math::zeros(3, 3);
        R(0, 0) = -1.; R(2, 1) = -1.; R(1, 2) = -1.;
```
I mean, it can be calculated as below, and if deg = 90, the result od_l was equal to the code. But why in my understanding it was 270 degree or -90 degree.
```c++
      yarp::sig::Matrix Rl = yarp::math::zeros(3, 3);
      Matrix Rx = yarp::math::zeros(3, 3);

      Rl(0, 0) = -1; Rl(1, 1) = -1; Rl(2, 2) = 1;  // original xyz of left hand

      int deg = 270;
      Rx(0, 0) = 1;
      Rx(1, 1) =  cos(deg/180*M_PI); Rx(1, 2) = -sin(deg/180*M_PI);
      Rx(2, 1) =  sin(deg/180*M_PI); Rx(2, 2) = cos(deg/180*M_PI);
      od_l = dcm2axis(Rx * Rl);
```
![image](https://github.com/pattacini/study-air-hockey/assets/7019692/75560388-0e9a-42bb-b8e1-81936f38b68a)

![2023-07-19_20-08](https://github.com/pattacini/study-air-hockey/assets/7019692/9e90d0fa-9823-40db-984a-3779ccf88ef7)

**Comment by pattacini** [comment]
>Does the initial pose(below) of study-1.cpp, generated from rotating the original xyz of left hand around its x axis by 270 or 90 degree?

Nope, it's a rotation of $`\pi`$ around the following axis:

$$
\left[0, -\frac{\sqrt{2}}{2}, \frac{\sqrt{2}}{2} \right]
$$

You can verify that by recruiting the function [**`dcm2axis`**](https://yarp.it/latest/namespaceyarp_1_1math.html#aefc08ebf2ccf238ed811b90f563b18c8):
```c++
auto axang = yarp::math::dcm2axis(R);
```

**Reply by jimingda** [reply]
What is the purpose of the following assignment, why not just assign like " y_max = 0.15; y_delta = 0.005; " ?
```
        y_max = std::abs(rf.check("y-max", yarp::os::Value(.15)).asFloat64());
        y_delta = std::abs(rf.check("y-delta", yarp::os::Value(.005)).asFloat64());
```

**Reply by pattacini** [reply]
`rf.check()` takes the values from the command line options, if specified. In case they're not, it falls back to default values, `.15` and `.005`, respectively.

**Comment by jimingda** [comment]
And my question is how to get the below xyz bases of left hand pose on study-1, my understanding is that it was left multiply by a rotate matrix on `R(0, 0) = -1; R(1, 1) = -1; R(2, 2) = 1;` I wander what is that rotate matrix
```
        yarp::sig::Matrix R = yarp::math::zeros(3, 3);
        R(0, 0) = -1.; R(2, 1) = -1.; R(1, 2) = -1.;
```

**Reply by pattacini** [reply, marked as answer]
The goal is to find the orientation of the left hand as shown in the figure below with respect to the root frame.

<img width="300" alt="image" src="https://github.com/pattacini/study-air-hockey/assets/3738070/9b1b7dfe-4a6b-4d55-bea1-91f5a5ea1cb2">

The left hand has the axes attached as you illustrated above, whereas the root frame is attached to the robot's waist with the x-axis pointing backward, the y-axis pointing rightward, and the z-axis pointing upward. See https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-forward-kinematics/icub-forward-kinematics/.

To achieve this, the most intuitive way is to come up with the direct cosine matrix that describes how a set of axes is oriented with respect to another.

In this context, we have:

| | hand x-axis | hand y-axis | hand z-axis |
| :---: | :---: | :---: | :---: |
| root x coordinates | -1 |  0 |  0 |
| root y coordinates |  0 |  0 | -1 |
| root z coordinates |  0 | -1 |  0 |

This matrix reads as follows:
- The first column contains the coordinates of the x-axis attached to the left hand given wrt the root frame. Indeed, the x-axis of the hand points counter root x-axis.
- The second and the third columns account for the y-axis and the z-axis of the hand, respectively.

Hope this helps clarify.

**Reply by pattacini** [reply]
Hi @jimingda 

Everything clear? May I close this discussion?

**Reply by jimingda** [reply]
Thanks. Very clear and i get insight understanding of direct cosine matrix.
